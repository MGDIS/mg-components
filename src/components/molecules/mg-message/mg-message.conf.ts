/**
 * List of all possibles message variants
 */
export const variants: string[] = ['info', 'warning', 'success', 'danger'];
