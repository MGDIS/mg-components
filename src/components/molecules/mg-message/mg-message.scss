@mixin message-variants($variants...) {
  @each $variant in $variants {
    &.mg-message--#{$variant} {
      &::before {
        background-color: hsl(var(--color-#{$variant}));
      }
      .mg-message__icon {
        color: hsl(var(--color-#{$variant}));
      }
    }
  }
}

.mg-message {
  // Display
  display: inline-block;
  position: relative;
  box-sizing: border-box;
  min-height: var(--default-size);
  overflow: hidden;
  // Decoration
  background-color: white;
  border-radius: var(--mg-message-border-radius);
  box-shadow: 0 0.3rem 0.8rem 0 hsla(var(--color-dark), 0.1);

  &::before {
    content: '';
    position: absolute;
    top: 0;
    left: 0;
    width: 0.3rem;
    height: 100%;
  }

  @include message-variants('info', 'warning', 'success', 'danger');

  &.mg-message--close-button {
    padding-right: 3.4rem;
  }

  &.mg-message--hide {
    display: none;
  }

  ::slotted(*) {
    // Removing default marging to slotted element
    margin: 0;
    padding: 0;
  }
}

.mg-message__icon {
  // Display
  position: absolute;
  top: calc((var(--default-size) - var(--mg-icon-regular-size)) / 2);
  left: 1.3rem;
  // Font
  line-height: 1;
}

.mg-message__content {
  display: flex;
  flex-wrap: wrap;
  justify-content: flex-end;
  align-content: stretch;
  padding: 0 1rem 0 3.8rem;
}

.mg-message__content-slot {
  flex-grow: 1;
  margin: 0.84rem 0;
}

.mg-message__content-separator {
  display: inline-block;
  width: 4.5rem;
  height: 0;
}

.mg-message__content-actions-slot {
  padding: 1rem 0;
  text-align: right;
}

.mg-message__close-button {
  position: absolute;
  top: 0;
  right: 0;
}

.mg-message,
mg-message {
  max-width: 100%;
}
