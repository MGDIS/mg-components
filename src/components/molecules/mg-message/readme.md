## Usage

### Types

The element can be of four different types depending on the message to be indicated:

- validation (green)
- information (blue)
- warning (orange)
- error (red)

When a trigger (button...) is set up, it can trigger the display of a message. By default, the execution of the trigger erases the previous messages in the screen of the same type.

## Specs

![](./mg-message/doc/img/mg-message-specs.png)

## CSS Variables

If needed some [variables](./?path=/story/css-variables--page) are available to customize the component:

- `--mg-message-border-radius`: Define message border radius, default: `0.3rem`

<!-- Auto Generated Below -->


## Properties

| Property      | Attribute      | Description                                                                                                                   | Type      | Default                  |
| ------------- | -------------- | ----------------------------------------------------------------------------------------------------------------------------- | --------- | ------------------------ |
| `closeButton` | `close-button` | Define if message has a cross button RG 01: https://jira.mgdis.fr/browse/PDA9-140                                             | `boolean` | `false`                  |
| `delay`       | `delay`        | Add a delay to hide/close message when it passed Value is defined in seconds and must greater than 2 seconds (PDA9-314 RG-06) | `number`  | `undefined`              |
| `hide`        | `hide`         | Define if message is hidden                                                                                                   | `boolean` | `false`                  |
| `identifier`  | `identifier`   | Identifier is used for the element ID (id is a reserved prop in Stencil.js) If not set, it will be created.                   | `string`  | `createID('mg-message')` |
| `variant`     | `variant`      | Message variant                                                                                                               | `string`  | `variants[0]`            |


## Events

| Event            | Description                           | Type                  |
| ---------------- | ------------------------------------- | --------------------- |
| `component-hide` | Emited event when message is hidden   | `CustomEvent<string>` |
| `component-show` | Emited event when message is diplayed | `CustomEvent<string>` |


## Dependencies

### Depends on

- [mg-card](../../atoms/mg-card)
- [mg-icon](../../atoms/mg-icon)
- [mg-button](../../atoms/mg-button)

### Graph
```mermaid
graph TD;
  mg-message --> mg-card
  mg-message --> mg-icon
  mg-message --> mg-button
  mg-button --> mg-icon
  style mg-message fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
