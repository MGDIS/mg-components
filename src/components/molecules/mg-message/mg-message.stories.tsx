import { h } from '@stencil/core';
import { variants } from './mg-message.conf';

export default {
  component: 'mg-message',
  title: 'Molecules/mg-message',
  argTypes: {
    variant: {
      options: variants,
      control: { type: 'select' },
    },
  },
};

/**
 * 1. camelCase arguments must be written in the template, for exemple labelOnTop must be placed in the template as label-on-top={args.labelOnTop}
 * 2. boolean arguments with a default true value must be added like display-character-left={args.displayCharacterLeft ? 'true' : 'false'}
 */

const Template = args => {
  // Extract slot so it won't be render as an attribute
  const slotContent = args.slotContent;
  delete args.slotContent;
  const slotActions = args.slotActions;
  delete args.slotActions;
  const closeButton = args.closeButton;
  delete args.closeButton;
  // return element
  return (
    <mg-message {...args} close-button={closeButton}>
      {slotContent && <span innerHTML={slotContent}></span>}
      {slotActions && <span slot="actions" innerHTML={slotActions}></span>}
    </mg-message>
  );
};

export const MgMessage = Template.bind({});

MgMessage.args = {
  slotContent: `<p>Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.</p>`,
  slotActions: ``,
  identifier: 'identifier',
  variant: variants[0], // info
  closeButton: false,
  hide: false,
};

export const WithCloseButton = Template.bind({});

WithCloseButton.args = {
  ...MgMessage.args,
  variant: 'danger',
  closeButton: true,
};

export const WithActions = Template.bind({});

WithActions.args = {
  ...MgMessage.args,
  variant: 'warning',
  slotActions: `<div slot="actions" class="mg-group-elements mg-group-elements--align-right"><mg-button>Primary</mg-button><mg-button variant="secondary">Secondary</mg-button></div>`,
};
