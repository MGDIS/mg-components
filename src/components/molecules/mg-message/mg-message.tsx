import { Component, Element, h, Prop, State, Watch } from '@stencil/core';
import { createID, ClassList } from '../../../utils/components.utils';
import { variants } from './mg-message.conf';
import { messages } from '../../../locales';

@Component({
  tag: 'mg-message',
  styleUrl: 'mg-message.scss',
  shadow: true,
})
export class MgMessage {

  /************
   * Internal *
   ************/

  private closeButtonId = "";

  /**************
   * Decorators *
   **************/

  @Element() element: HTMLMgMessageElement;

  /**
   * Identifier is used for the element ID (id is a reserved prop in Stencil.js)
   * If not set, it will be created.
   */
  @Prop() identifier?: string = createID('mg-message');

  /**
   * Message variant
   */
  @Prop() variant?: string = variants[0];
  @Watch('variant')
  validateType(newValue: string) {
    if(!variants.includes(newValue)) {
      throw new Error(`<mg-message> prop "variant" must be one of : ${variants.join(', ')}`);
    }
    this.classList.add(`mg-message--${this.variant}`);
  }

  /**
   * Define if message has a cross button
   * RG 01: https://jira.mgdis.fr/browse/PDA9-140
   */
  @Prop({ mutable: true }) closeButton?: boolean = false;
  @Watch('closeButton')
  validateCloseButton(newValue: boolean) {
    if(newValue && this.hasActions) {
      this.closeButton = false;
      throw new Error('<mg-message> prop "close-button" can\'t be used with the actions slot.')
    }
  }

  /**
   * Define if message is hidden
   */
   @Prop({ mutable: true, reflect: true }) hide?: boolean = false;
   @Watch('hide')
   validateHide(newValue: boolean) {
     if(newValue) this.classList.add('mg-message--hide')
     else this.classList.delete('mg-message--hide')
   }

  /**
   * Component classes
   */
  @State() classList: ClassList = new ClassList(['mg-message']);

  /**
   * Define if component is using actions slot
   */
  @State() hasActions: boolean = false;

  /**
   * Handle close button
   */
  private handleClose = () => {
    this.hide = true;
  }

  /**
   * Get icon corresponding to variant
   */
  private getIcon = ():string => {
    switch (this.variant) {
      case "info":
        return "info-circle"
      case "warning":
        return "exclamation-triangle"
      case "success":
          return "check-circle"
      case "danger":
          return "exclamation-circle"
      default:
        break;
    }
  }

  /*************
   * Lifecycle *
   *************/

  /**
   * Check if component props are well configured on init
   */
   componentWillLoad() {
    this.validateType(this.variant);
    // Check if close button is an can be activated
    this.hasActions = this.element.querySelector('[slot="actions"]') !== null;
    this.validateCloseButton(this.closeButton);
    if(this.closeButton) {
      this.classList.add('mg-message--close-button');
      this.closeButtonId = `${this.identifier}-close-button`;
    }
    this.validateHide(this.hide);
  }

  render() {
    return (
      <div id={this.identifier} class={this.classList.join()} role={this.variant === "info" ? "status" : "alert"}>
        <span class="mg-message__icon">
          <mg-icon icon={this.getIcon()}></mg-icon>
        </span>
        <div class="mg-message__content">
          <span class="mg-message__content__slot">
            <slot></slot>
          </span>
          { this.hasActions && <span class="mg-message__content__separator"></span> }
          { this.hasActions && <span class="mg-message__content__actions-slot">
              <slot name="actions"></slot>
            </span>
          }
        </div>
        { this.closeButton && <span class="mg-message__close-button">
            <mg-button identifier={this.closeButtonId} is-icon variant="flat" label={messages.message.closeButton} onClick={this.handleClose}>
              <mg-icon icon="cross"></mg-icon>
            </mg-button>
          </span>
        }
      </div>
    );
  }

}
