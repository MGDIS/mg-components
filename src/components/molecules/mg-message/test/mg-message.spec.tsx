import { h } from '@stencil/core';
import { newSpecPage } from '@stencil/core/testing';
import { MgMessage } from '../mg-message';
import { variants } from '../mg-message.conf';

const getDefaultContent = () => (
  <p>
    <strong>Strong</strong> content!
  </p>
);

const getPage = (args, content) =>
  newSpecPage({
    components: [MgMessage],
    template: () => <mg-message {...args}>{content}</mg-message>,
  });

describe('mg-message', () => {
  describe.each(variants)('Should render a %s message', variant => {
    test.each([
      { identifier: 'identifier' },
      { identifier: 'identifier', closeButton: true },
      { identifier: 'identifier', closeButton: true, lang: 'fr' },
      { identifier: 'identifier', closeButton: true, lang: 'xx' },
    ])('with args %s', async args => {
      const { root } = await getPage({ ...args, variant }, getDefaultContent());
      expect(root).toMatchSnapshot();
    });
  });

  test('Should replace classes on variant changes', async () => {
    const page = await getPage({ identifier: 'identifier' }, getDefaultContent());
    const element = page.doc.querySelector('mg-message');
    let classInfo = element.shadowRoot.querySelector('.mg-message--info');

    expect(classInfo).not.toBeNull();

    element.variant = 'danger';
    await page.waitForChanges();

    classInfo = element.shadowRoot.querySelector('.mg-message--info');
    const classDanger = element.shadowRoot.querySelector('.mg-message--danger');

    expect(classInfo).toBeNull();
    expect(classDanger).not.toBeNull();
  });

  test.each(['', 'blu'])('Should throw error with invalid variant property : %s', async variant => {
    expect.assertions(1);
    try {
      await getPage({ identifier: 'identifier', variant }, getDefaultContent());
    } catch (err) {
      expect(err.message).toMatch('<mg-message> prop "variant" must be one of :');
    }
  });

  test('Should throw error with invalid delay property : 50ms', async () => {
    expect.assertions(1);
    try {
      await getPage({ identifier: 'identifier', delay: 1 }, getDefaultContent());
    } catch (err) {
      expect(err.message).toMatch('<mg-message> prop "delay" must be greater than 2 seconds.');
    }
  });

  test('Should throw error when using prop "close-button" with an action slot', async () => {
    expect.assertions(1);
    try {
      await getPage({ identifier: 'identifier', closeButton: true }, [
        getDefaultContent(),
        <div slot="actions" class="mg-group-elements mg-group-elements--align-right">
          <mg-button>Primary</mg-button>
          <mg-button variant="secondary">Secondary</mg-button>
        </div>,
      ]);
    } catch (err) {
      expect(err.message).toMatch('<mg-message> prop "close-button" can\'t be used with the actions slot.');
    }
  });

  test('Should hide message on close button', async () => {
    const args = { identifier: 'identifier', closeButton: true };
    const page = await getPage(args, getDefaultContent());

    const element = page.doc.querySelector('mg-message');
    const button = element.shadowRoot.querySelector('mg-button');

    jest.spyOn(page.rootInstance.componentHide, 'emit');

    button.dispatchEvent(new CustomEvent('click', { bubbles: true }));
    await page.waitForChanges();

    expect(page.rootInstance.componentHide.emit).toHaveBeenCalledTimes(1);
    expect(page.rootInstance.classList.join()).toContain('mg-message--hide');
  });

  test('Should hide message on delay', async () => {
    const args = { identifier: 'identifier', delay: 2 };
    const page = await getPage(args, getDefaultContent());

    jest.spyOn(page.rootInstance.componentHide, 'emit');
    jest.spyOn(page.rootInstance.componentShow, 'emit');

    expect(page.rootInstance.classList.join()).not.toContain('mg-message--hide');

    jest.advanceTimersByTime(2000);

    expect(page.rootInstance.classList.join()).toContain('mg-message--hide');
    expect(page.rootInstance.componentHide.emit).toHaveBeenCalledTimes(1);

    page.rootInstance.hide = false;
    await page.waitForChanges();

    expect(page.rootInstance.classList.join()).not.toContain('mg-message--hide');
    expect(page.rootInstance.componentShow.emit).toHaveBeenCalledTimes(1);

    jest.advanceTimersByTime(2000);

    expect(page.rootInstance.classList.join()).toContain('mg-message--hide');
    expect(page.rootInstance.componentHide.emit).toHaveBeenCalledTimes(2);
  });
});
