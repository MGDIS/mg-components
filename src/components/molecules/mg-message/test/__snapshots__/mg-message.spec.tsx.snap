// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mg-message Should render a danger message with args { identifier: 'identifier' } 1`] = `
<mg-message>
  <mock:shadow-root>
    <div class="mg-message mg-message--danger" id="identifier" role="alert">
      <span class="mg-message__icon">
        <mg-icon icon="danger"></mg-icon>
      </span>
      <div class="mg-message__content">
        <span class="mg-message__content__slot">
          <slot></slot>
        </span>
      </div>
    </div>
  </mock:shadow-root>
  <p>
    <strong>
      Strong
    </strong>
    content!
  </p>
</mg-message>
`;

exports[`mg-message Should render a danger message with args { identifier: 'identifier', closeButton: true } 1`] = `
<mg-message>
  <mock:shadow-root>
    <div class="mg-message mg-message--close-button mg-message--danger" id="identifier" role="alert">
      <span class="mg-message__icon">
        <mg-icon icon="danger"></mg-icon>
      </span>
      <div class="mg-message__content">
        <span class="mg-message__content__slot">
          <slot></slot>
        </span>
      </div>
      <span class="mg-message__close-button">
        <mg-button identifier="identifier-close-button" is-icon="" label="Fermer le message" variant="flat">
          <mg-icon icon="cross" size="small"></mg-icon>
        </mg-button>
      </span>
    </div>
  </mock:shadow-root>
  <p>
    <strong>
      Strong
    </strong>
    content!
  </p>
</mg-message>
`;

exports[`mg-message Should render a info message with args { identifier: 'identifier' } 1`] = `
<mg-message>
  <mock:shadow-root>
    <div class="mg-message mg-message--info" id="identifier" role="status">
      <span class="mg-message__icon">
        <mg-icon icon="info"></mg-icon>
      </span>
      <div class="mg-message__content">
        <span class="mg-message__content__slot">
          <slot></slot>
        </span>
      </div>
    </div>
  </mock:shadow-root>
  <p>
    <strong>
      Strong
    </strong>
    content!
  </p>
</mg-message>
`;

exports[`mg-message Should render a info message with args { identifier: 'identifier', closeButton: true } 1`] = `
<mg-message>
  <mock:shadow-root>
    <div class="mg-message mg-message--close-button mg-message--info" id="identifier" role="status">
      <span class="mg-message__icon">
        <mg-icon icon="info"></mg-icon>
      </span>
      <div class="mg-message__content">
        <span class="mg-message__content__slot">
          <slot></slot>
        </span>
      </div>
      <span class="mg-message__close-button">
        <mg-button identifier="identifier-close-button" is-icon="" label="Fermer le message" variant="flat">
          <mg-icon icon="cross" size="small"></mg-icon>
        </mg-button>
      </span>
    </div>
  </mock:shadow-root>
  <p>
    <strong>
      Strong
    </strong>
    content!
  </p>
</mg-message>
`;

exports[`mg-message Should render a success message with args { identifier: 'identifier' } 1`] = `
<mg-message>
  <mock:shadow-root>
    <div class="mg-message mg-message--success" id="identifier" role="alert">
      <span class="mg-message__icon">
        <mg-icon icon="success"></mg-icon>
      </span>
      <div class="mg-message__content">
        <span class="mg-message__content__slot">
          <slot></slot>
        </span>
      </div>
    </div>
  </mock:shadow-root>
  <p>
    <strong>
      Strong
    </strong>
    content!
  </p>
</mg-message>
`;

exports[`mg-message Should render a success message with args { identifier: 'identifier', closeButton: true } 1`] = `
<mg-message>
  <mock:shadow-root>
    <div class="mg-message mg-message--close-button mg-message--success" id="identifier" role="alert">
      <span class="mg-message__icon">
        <mg-icon icon="success"></mg-icon>
      </span>
      <div class="mg-message__content">
        <span class="mg-message__content__slot">
          <slot></slot>
        </span>
      </div>
      <span class="mg-message__close-button">
        <mg-button identifier="identifier-close-button" is-icon="" label="Fermer le message" variant="flat">
          <mg-icon icon="cross" size="small"></mg-icon>
        </mg-button>
      </span>
    </div>
  </mock:shadow-root>
  <p>
    <strong>
      Strong
    </strong>
    content!
  </p>
</mg-message>
`;

exports[`mg-message Should render a warning message with args { identifier: 'identifier' } 1`] = `
<mg-message>
  <mock:shadow-root>
    <div class="mg-message mg-message--warning" id="identifier" role="alert">
      <span class="mg-message__icon">
        <mg-icon icon="warning"></mg-icon>
      </span>
      <div class="mg-message__content">
        <span class="mg-message__content__slot">
          <slot></slot>
        </span>
      </div>
    </div>
  </mock:shadow-root>
  <p>
    <strong>
      Strong
    </strong>
    content!
  </p>
</mg-message>
`;

exports[`mg-message Should render a warning message with args { identifier: 'identifier', closeButton: true } 1`] = `
<mg-message>
  <mock:shadow-root>
    <div class="mg-message mg-message--close-button mg-message--warning" id="identifier" role="alert">
      <span class="mg-message__icon">
        <mg-icon icon="warning"></mg-icon>
      </span>
      <div class="mg-message__content">
        <span class="mg-message__content__slot">
          <slot></slot>
        </span>
      </div>
      <span class="mg-message__close-button">
        <mg-button identifier="identifier-close-button" is-icon="" label="Fermer le message" variant="flat">
          <mg-icon icon="cross" size="small"></mg-icon>
        </mg-button>
      </span>
    </div>
  </mock:shadow-root>
  <p>
    <strong>
      Strong
    </strong>
    content!
  </p>
</mg-message>
`;
